packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp

```diff
@@ -895,6 +1080,687 @@ void Add(
+//  <Make CrowptrTemp into prefix sum array>
+//  <Allocate oversized Cvals and Ccolind>
+//2: just copy scaled a,b values into oversized C values/colind (filling it completely)
+//  Note: these values aren't condensed or sorted yet (so, may be duplicate colinds, but that's expected)
+//  <Allocate Crowptr>
+//3: Sort each row's colinds (permuting values at same time), then count unique colinds (write that to Crowptr(i))
+//  <Make Crowptr into prefix sum array>
+//  <Allocate correctly sized Cvals and Ccolind>
+//4: Condense values from oversized Cvals/Ccolind to correctly sized
+//  <done: have everything needed to build final C>
+
+template<typename Ordinal, typename Values, typename RowPtrs, typename ColInds>
+struct EntryUpperBoundFunctor
+{
+  EntryUpperBoundFunctor(const Values& Avals_, const RowPtrs& Arowptrs_, const ColInds& Acolinds_, const Values& Bvals_, const RowPtrs& Browptrs_, const ColInds& Bcolinds_, RowPtrs& CrowcountsTemp_) :
+  Avals(Avals_), Arowptrs(Arowptrs_), Acolinds(Acolinds_),
+  Bvals(Bvals_), Browptrs(Browptrs_), Bcolinds(Bcolinds_),
+  CrowcountsTemp(CrowcountsTemp_) {}
+  KOKKOS_INLINE_FUNCTION void operator()(const size_t i) const
+  {
+    CrowcountsTemp(i) = (Arowptrs(i + 1) - Arowptrs(i)) + (Browptrs(i + 1) - Browptrs(i));
+  }
+  const Values& Avals;
+  const RowPtrs& Arowptrs;
+  const ColInds& Acolinds;
+  const Values& Bvals;
+  const RowPtrs& Browptrs;
+  const ColInds& Bcolinds;
+  RowPtrs& CrowcountsTemp;
+};
+
+template<typename Scalar, typename Ordinal, typename Values, typename RowPtrs, typename ColInds>
+struct UnmergedSumFunctor
+{
+  UnmergedSumFunctor(const Values& Avals_, const RowPtrs& Arowptrs_, const ColInds& Acolinds_, const Scalar& scalarA_,
+           const Values& Bvals_, const RowPtrs& Browptrs_, const ColInds& Bcolinds_, const Scalar& scalarB_,
+           Values& CvalsOver_, const RowPtrs& CrowptrsTemp_, ColInds& CcolindsOver_) :
+    Avals(Avals_), Arowptrs(Arowptrs_), Acolinds(Acolinds_), scalarA(scalarA_),
+    Bvals(Bvals_), Browptrs(Browptrs_), Bcolinds(Bcolinds_), scalarB(scalarB_),
+    CvalsOver(CvalsOver_), CrowptrsTemp(CrowptrsTemp_), CcolindsOver(CcolindsOver_)
+  {}
+  KOKKOS_INLINE_FUNCTION void operator()(const size_t i) const
+  {
+    Ordinal inserted = 0;
+    auto rowStart = CrowptrsTemp(i);
+    auto arowstart = Arowptrs(i);
+    auto arowlen = Arowptrs(i + 1) - arowstart;
+    auto browstart = Browptrs(i);
+    auto browlen = Browptrs(i + 1) - browstart;
+    for(Ordinal j = 0; j < arowlen; j++)
+    {
+      CvalsOver(rowStart + inserted) = scalarA * Avals(arowstart + j);
+      CcolindsOver(rowStart + inserted) = Acolinds(arowstart + j);
+      inserted++;
+    }
+    for(Ordinal j = 0; j < browlen; j++)
+    {
+      CvalsOver(rowStart + inserted) = scalarB * Bvals(browstart + j);
+      CcolindsOver(rowStart + inserted) = Bcolinds(browstart + j);
+      inserted++;
+    }
+  }
+  const Values& Avals;
```

Comment:
See above note on Kokkos::View needing to be stored by value in functors, not by reference.